crates/bevy_core/src/time/timer.rs

```diff
@@ -27,15 +27,15 @@ impl Timer {
     }
 
     pub fn tick(&mut self, delta: f32) {
-        self.elapsed = (self.elapsed + delta).min(self.duration);
+        self.elapsed = self.elapsed + delta;
         if self.elapsed >= self.duration {
             self.finished = true;
         }
     }
 
     pub fn reset(&mut self) {
         self.finished = false;
-        self.elapsed = 0.0;
+        self.elapsed = self.elapsed % self.duration;
```

Comment:
I think this changes the semantics around `reset` in a way that would affect other valid use cases. Consider when the intention is to reset a timer that has **not** finished yet. In that case I think you would want it to return to zero.

Reply:
Maybe another adding a new method is called for with a name like: `resume`, `continue` or `rollover` ?

Reply:
Tested it with a resettable timer and you're totally right! Fixed in the next commit.

Reply:
Not sure what the added value would be of an extra method. Both cases can be covered by a single function as far as I can tell.

Reply:
I suggested a new method because `reset` to me, sounds like it should always zero out the timer.

The additional change you added to handle reset before a timer is `finished` improves the situation but I think there are still situations where this automatic carry over would be surprising.

What about a timer that is only reset in order to measure an infrequent event (not looping all the time). Imagine a timer used to leave an informational message on the screen for a specified interval. It would be surprising if the timer carried over the previous elapsed time when it was reset. I suppose if it were an infrequent timer like that, it wouldn't have a significant effect if the timer was only carried forward by a single run loop's worth of elapsed time. But as it's currently implemented it my go along ticking the elapsed variable for a long duration and by the time reset is called it's at some arbitrary value.

That made me think a new `method` or a new type like LoopingTimer might be nice.

If we want to make this support both looping with cary over, as well as infrequent event time measurements I think a little bit more logic in `tick()` might get us there.

How about something like this:
```rust
//...
    pub fn tick(&mut self, delta: f32) {
        if (self.finished) {
            // A tick after already finished before `reset` was called indicates that carry over
            // of elapsed time for a looping timer was not intended.
            self.elapsed = self.duration;
            return;
        }
        self.elapsed = self.elapsed + delta;
        if self.elapsed >= self.duration {
            self.finished = true;
        }
    }

    pub fn reset(&mut self) {
        self.elapsed = if self.finished {
            self.elapsed % self.duration
        } else {
            0.0
        };
        self.finished = false;
    }
//...
```

Note: I haven't had a chance to test this code yet. :)

Reply:
Here are some tests for the above:

```rust
#[cfg(test)]
mod tests {
    use super::{Timer};

    #[test]
    fn test_carry_over() {
        let mut timer = Timer::from_seconds(3.0);
        timer.tick(3.5);
        assert!(timer.finished == true);
        timer.reset();
        // Carry over because the timer was finished during the most recent tick.
        assert_eq!(timer.elapsed, 0.5);
    }

    #[test]
    fn test_carry_over_not_applied_to_unfinished_timer() {
        let mut timer = Timer::from_seconds(3.0);
        timer.tick(2.5);
        assert!(timer.finished == false);
        timer.reset();
        // No carry over since timer wasn't finished.
        assert_eq!(timer.elapsed, 0.0);
    }

    #[test]
    fn test_carry_over_not_applied_to_rarely_reset_timer() {
        let mut timer = Timer::from_seconds(3.0);
        timer.tick(2.5);
        assert!(timer.finished == false);
        timer.tick(2.5);
        assert!(timer.finished == true);
        assert_eq!(timer.elapsed, 5.0);

        // Tick again after finished without reseting means the timer is set to `duration`
        timer.tick(2.5);
        assert_eq!(timer.elapsed, 3.0);

        timer.reset();
        // No carry over since timer wasn't finished after the most recent tick.
        assert_eq!(timer.elapsed, 0.0);
    }
}

```

```
running 3 tests
test tests::test_carry_over_not_applied_to_unfinished_timer ... ok
test tests::test_carry_over_not_applied_to_rarely_reset_timer ... ok
test tests::test_carry_over ... ok
```

Reply:
Thanks for the explanation! That does clear it up a lot for me.

Given the example used in the introduction (https://bevyengine.org/learn/book/getting-started/resources/) I had the impression `Timer` was supposed to be a repeating construct from the get go, but the case you've pointed out makes a lot of sense too.

Maybe adding a `repeating` flag in the constructors would give the most intuitive behavior. In non repeating mode the timer stops after finishing once, then `reset()` should be called to re-use the timer. In repeating mode the timer sets the finished state during the update in which the duration was reached and automatically resets it's elapsed time carrying over any excess. In the next update `finished` should then be false again. This also seems the most ergonimical since for repeating timers you don't have to keep calling a reset/resume function manually; you just set it to go and keep check of whether or not it finished.

Reply:
As an example of the above, I've yet to actually test this:
<details>

```rust
    pub fn tick(&mut self, delta: f32) {
        match (self.repeating, self.elapsed >= self.duration) {
            // non repeating timer
            (false, false) => {
                self.elapsed = self.elapsed + delta;
            },
            (false, true) => {
                self.finished = true;
            },

            // repeating timer
            (true, false) => {
                self.elapsed = self.elapsed + delta;
                self.finished = false;
            }
            (true, true) => {
                self.elapsed = self.elapsed % self.duration;
                self.finished = true;
            }
        }
    }

    pub fn reset(&mut self) {
        self.finished = false;
        self.elapsed = 0.0;
    }
```

</details>